File: Algorithm/Longest Common Subsequence/JAVA/LCS.java
public class LCS {
   private int length;
   private Object object;
   private LCS prev;

   private LCS() {
      length = 0;
      object = null;
      prev = null;
   }

   public static Object[] find(Object[] seq1, Object[] seq2) {
      LCS element;
      LCS[][] table = new LCS[seq1.length + 1][seq2.length + 1];

      for (int i = 0; i < table.length; i++) {
         for (int j = 0; j < table[i].length; j++) {
            table[i][j] = new LCS();
         }
      }

      for (int i = 0; i < seq1.length; i++) {
         for (int j = 0; j < seq2.length; j++) {
            if (seq1[i].equals(seq2[j])) {
               table[i+1][j+1].length = table[i][j].length + 1;
               table[i+1][j+1].object = seq1[i];
               table[i+1][j+1].prev = table[i][j];
            } else {
               if (table[i][j+1].length > table[i+1][j].length) {
                  table[i+1][j+1].length = table[i][j+1].length;
                  table[i+1][j+1].prev = table[i][j+1];
               } else {
                  table[i+1][j+1].length = table[i+1][j].length;
                  table[i+1][j+1].prev = table[i+1][j];
               }
            }
         }
      }

      element = table[seq1.length][seq2.length];
      Object[] lcs = new Object[element.length];
      int index = element.length;
      while (element != null) {
         if (element.object != null) {
            --index;
            lcs[index] = element.object;
         }
         element = element.prev;
      }

      return lcs;
   }
}

File: Algorithm/Longest Common Subsequence/Python/lcs.py
def LCS(seq1, seq2, equal):
   s1 = range(0, len(seq1)+1)
   s2 = range(0, len(seq2)+1)
   score = [[0 for s in s2] for s in s1]
   prev = [[0 for s in s2] for s in s1]
   for i in s1[:-1]:
      for j in s2[:-1]:
         if equal(seq1[i], seq2[j]):
            score[i+1][j+1] = score[i][j] + 1
         else:
            if score[i][j+1] > score[i+1][j]:
               score[i+1][j+1] = score[i][j+1]
               prev[i+1][j+1] = 1
            else:
               score[i+1][j+1] = score[i+1][j]
               prev[i+1][j+1] = -1
   seq = [None for s in range(0, score[-1][-1])]
   i = len(seq1)
   j = len(seq2)
   while i != 0 and j != 0:
      if prev[i][j] == 0:
         seq[score[i][j]-1] = seq1[i-1]
         i -= 1
         j -= 1
      elif prev[i][j] == 1:
         i -= 1
      else:
         j -= 1

   return seq
